apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs

```diff
@@ -0,0 +1,120 @@
+    /// <param name="cancellationToken">A cancellation token for the operation.</param>
+    /// <returns>A <see cref="AggregateQuerySnapshot"/> contains results of the Aggregate functions.</returns>
+    public Task<AggregateQuerySnapshot> GetSnapshotAsync(CancellationToken cancellationToken = default) =>
+        GetSnapshotAsync(null, cancellationToken);
+
+    internal async Task<AggregateQuerySnapshot> GetSnapshotAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        IAsyncEnumerable<RunAggregationQueryResponse> responseStream = GetAggregationQueryResponseStreamAsync(transactionId, cancellationToken);
+        Timestamp? readTime = null;
+        Value count = null;
+        await responseStream.ForEachAsync(response =>
+        {
+            if (response.Result.AggregateFields != null)
+            {
+                response.Result.AggregateFields.TryGetValue(CountAlias, out count);
+            }
+            if (readTime == null && response.ReadTime != null)
+            {
+                readTime = Timestamp.FromProto(response.ReadTime);
+            }
+        }, cancellationToken).ConfigureAwait(false);
+
+        GaxPreconditions.CheckState(readTime != null, "The stream returned from RunAggregationQuery did not provide a read timestamp.");
+        var snapshot = new AggregateQuerySnapshot(this, readTime.Value, count?.IntegerValue ?? 0);
+        return snapshot;
+    }
+
+    private IAsyncEnumerable<RunAggregationQueryResponse> GetAggregationQueryResponseStreamAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        var structuredAggregationQuery = new StructuredAggregationQuery
+        {
+            StructuredQuery = _query.ToStructuredQuery(),
+            Aggregations = { new Aggregation { Alias = CountAlias, Count = new Count() } }
+        };
+
+        RunAggregationQueryRequest request = new RunAggregationQueryRequest
+        {
+            Parent = _query.ParentPath,
+            StructuredAggregationQuery = structuredAggregationQuery
+        };
+        if (transactionId != null)
+        {
+            request.Transaction = transactionId;
+        }
+        var settings = CallSettings.FromCancellationToken(cancellationToken);
+        var response = _query.Database.Client.RunAggregationQuery(request, settings);
+        return response.GetResponseStream();
+    }
+
+    /// <inheritdoc />
+    public override int GetHashCode() => _query.GetHashCode();
+
+    /// <summary> 
+    /// Determines whether <paramref name="other"/> is equal to this instance.
+    /// </summary>
+    /// <returns><c>true</c> if the specified object is equal to this instance; otherwise <c>false</c>.</returns>
+    public bool Equals(AggregateQuery other)
+    {
+        if (ReferenceEquals(other, this))
+        {
+            return true;
+        }
+        if (other is null)
+        {
+            return false;
+        }
+        if (GetType() != other.GetType())
```

Comment:
No need for this - the type is sealed.

Again though, I'd encourage you to include the aggregations as part of the state, both here and for GetHashCode.
If we represent "we've asked for count" as state, even though it's the only available aggregation available now, it'll make it much easier to add support for other aggregations later.

Reply:
Make Sense Jon will do.
Thanks